File: sdks/node/src/__tests__/webhooks.test.ts
import { Webhooks } from '../resources/webhooks';
import { ApiClient } from '../client';
import {
  ProviderEnum,
  SubscribeWebhookParams,
  SubscribeWebhookResponse,
  UnsubscribeWebhookParams,
  UnsubscribeWebhookResponse,
  ValidationError,
  AuthenticationError,
  ServerError,
} from '../types';

jest.mock('../client');

describe('Webhooks Resource', () => {
  let webhooks: Webhooks;
  let mockApiClient: jest.Mocked<ApiClient>;

  beforeEach(() => {
    mockApiClient = new ApiClient({
      clientId: 'id',
      clientSecret: 'secret',
      redirectUri: 'uri',
    }) as jest.Mocked<ApiClient>;

    webhooks = new Webhooks(mockApiClient);
  });

  describe('subscribeWebhook', () => {
    it('should call client.post with correct path and body', async () => {
      const params: SubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        calendarId: 'primary',
      };

      const mockResponse: SubscribeWebhookResponse = {
        success: true,
        provider: 'google',
        subscription: {
          channelId: 'channel-123',
          resourceId: 'resource-456',
          expiration: '2026-10-01T00:00:00.000Z',
        },
        serverWebhookUrl: 'https://connect.mobiscroll.com/api/webhook-callback',
        channelId: 'channel-123',
      };

      mockApiClient.post.mockResolvedValue({
        data: mockResponse,
        status: 200,
        headers: {},
      });

      const result = await webhooks.subscribeWebhook(params);

      expect(mockApiClient.post).toHaveBeenCalledWith('/subscribe-webhook', params);
      expect(result).toEqual(mockResponse);
    });

    it('should pass optional channelId and expiration through', async () => {
      const params: SubscribeWebhookParams = {
        provider: ProviderEnum.Microsoft,
        calendarId: 'AAMk==',
        channelId: 'custom-channel',
        expiration: 1893456000000,
      };

      const mockResponse: SubscribeWebhookResponse = {
        success: true,
        provider: 'microsoft',
        subscription: { channelId: 'custom-channel' },
        serverWebhookUrl: 'https://connect.mobiscroll.com/api/webhook-callback',
        channelId: 'custom-channel',
      };

      mockApiClient.post.mockResolvedValue({
        data: mockResponse,
        status: 200,
        headers: {},
      });

      const result = await webhooks.subscribeWebhook(params);

      expect(mockApiClient.post).toHaveBeenCalledWith('/subscribe-webhook', params);
      expect(result).toEqual(mockResponse);
    });

    it('should propagate a ValidationError for a missing provider/calendarId (400)', async () => {
      const params = { calendarId: 'primary' } as unknown as SubscribeWebhookParams;
      mockApiClient.post.mockRejectedValue(new ValidationError('provider is required'));

      await expect(webhooks.subscribeWebhook(params)).rejects.toThrow(ValidationError);
    });

    it('should propagate an AuthenticationError for a bad/missing bearer token (401)', async () => {
      const params: SubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        calendarId: 'primary',
      };
      mockApiClient.post.mockRejectedValue(new AuthenticationError('Invalid token'));

      await expect(webhooks.subscribeWebhook(params)).rejects.toThrow(AuthenticationError);
    });

    it('should propagate a ServerError (500)', async () => {
      const params: SubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        calendarId: 'primary',
      };
      mockApiClient.post.mockRejectedValue(new ServerError('Internal server error', 500));

      await expect(webhooks.subscribeWebhook(params)).rejects.toThrow(ServerError);
    });
  });

  describe('unsubscribeWebhook', () => {
    it('should call client.post with correct path and body', async () => {
      const params: UnsubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        channelId: 'channel-123',
        resourceId: 'resource-456',
      };

      const mockResponse: UnsubscribeWebhookResponse = { success: true };

      mockApiClient.post.mockResolvedValue({
        data: mockResponse,
        status: 200,
        headers: {},
      });

      const result = await webhooks.unsubscribeWebhook(params);

      expect(mockApiClient.post).toHaveBeenCalledWith('/unsubscribe-webhook', params);
      expect(result).toEqual(mockResponse);
    });

    it('should treat success: true with an explanatory message as final', async () => {
      const params: UnsubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        channelId: 'channel-123',
      };

      const mockResponse: UnsubscribeWebhookResponse = {
        success: true,
        message: 'Subscription had already expired upstream; local mapping removed.',
      };

      mockApiClient.post.mockResolvedValue({
        data: mockResponse,
        status: 200,
        headers: {},
      });

      const result = await webhooks.unsubscribeWebhook(params);

      expect(result.success).toBe(true);
      expect(result.message).toBeDefined();
    });

    it('should propagate a ValidationError for a missing provider/channelId (400)', async () => {
      const params = { provider: ProviderEnum.Google } as unknown as UnsubscribeWebhookParams;
      mockApiClient.post.mockRejectedValue(new ValidationError('channelId is required'));

      await expect(webhooks.unsubscribeWebhook(params)).rejects.toThrow(ValidationError);
    });

    it('should propagate an AuthenticationError for a bad/missing bearer token (401)', async () => {
      const params: UnsubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        channelId: 'channel-123',
      };
      mockApiClient.post.mockRejectedValue(new AuthenticationError('Invalid token'));

      await expect(webhooks.unsubscribeWebhook(params)).rejects.toThrow(AuthenticationError);
    });

    it('should propagate a ServerError (500)', async () => {
      const params: UnsubscribeWebhookParams = {
        provider: ProviderEnum.Google,
        channelId: 'channel-123',
      };
      mockApiClient.post.mockRejectedValue(new ServerError('Internal server error', 500));

      await expect(webhooks.unsubscribeWebhook(params)).rejects.toThrow(ServerError);
    });
  });
});

File: sdks/go/webhooks_test.go
package mobiscroll_test

import (
	"context"
	"errors"
	"strings"
	"testing"

	mobiscroll "github.com/acidb/mobiscroll-connect-sdks/sdks/go"
	"github.com/acidb/mobiscroll-connect-sdks/sdks/go/testsupport"
)

func TestWebhooks_SubscribeWebhookSendsJSONBodyAndParsesResponse(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.EnqueueJSON(`{
		"success": true,
		"provider": "google",
		"subscription": {"channelId": "chan1", "resourceId": "res1", "expiration": "2026-06-01T00:00:00.000Z"},
		"serverWebhookUrl": "https://connect.mobiscroll.com/api/webhook-callback",
		"channelId": "chan1"
	}`)

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at"})

	resp, err := c.Webhooks().SubscribeWebhook(context.Background(), &mobiscroll.SubscribeWebhookParams{
		Provider:   mobiscroll.ProviderGoogle,
		CalendarID: "primary",
		Expiration: mobiscroll.Ptr(int64(1780000000000)),
	})
	if err != nil {
		t.Fatalf("SubscribeWebhook: %v", err)
	}
	if !resp.Success || resp.Provider != "google" || resp.ChannelID != "chan1" {
		t.Fatalf("unexpected response: %+v", resp)
	}
	if resp.Subscription.ResourceID != "res1" || resp.Subscription.Expiration != "2026-06-01T00:00:00.000Z" {
		t.Fatalf("unexpected subscription: %+v", resp.Subscription)
	}
	if resp.ServerWebhookURL == "" {
		t.Fatalf("expected ServerWebhookURL to be set, got %+v", resp)
	}

	req := srv.Requests()[0]
	if req.Method != "POST" || req.Path != "/api/subscribe-webhook" {
		t.Fatalf("unexpected request: %s %s", req.Method, req.Path)
	}
	if !strings.HasPrefix(req.Header.Get("Content-Type"), "application/json") {
		t.Errorf("expected JSON content type, got %q", req.Header.Get("Content-Type"))
	}
	if got := req.Header.Get("Authorization"); got != "Bearer at" {
		t.Errorf("expected Bearer header, got %q", got)
	}
	body := string(req.Body)
	for _, want := range []string{`"provider":"google"`, `"calendarId":"primary"`, `"expiration":1780000000000`} {
		if !strings.Contains(body, want) {
			t.Errorf("body missing %q: %s", want, body)
		}
	}
	if strings.Contains(body, "channelId") {
		t.Errorf("expected channelId to be omitted when unset, got body: %s", body)
	}
}

func TestWebhooks_SubscribeWebhookNilParams(t *testing.T) {
	c := mobiscroll.NewClient("id", "secret", "https://app/cb")
	if _, err := c.Webhooks().SubscribeWebhook(context.Background(), nil); err == nil {
		t.Fatal("expected error for nil params")
	}
}

func TestWebhooks_SubscribeWebhookValidationError(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.Enqueue(testsupport.MockResponse{
		Status:  400,
		Body:    `{"message":"calendarId is required"}`,
		Headers: map[string]string{"Content-Type": "application/json"},
	})

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at"})

	_, err := c.Webhooks().SubscribeWebhook(context.Background(), &mobiscroll.SubscribeWebhookParams{
		Provider: mobiscroll.ProviderGoogle,
	})
	var ve *mobiscroll.ValidationError
	if !errors.As(err, &ve) {
		t.Fatalf("expected *ValidationError, got %T: %v", err, err)
	}
}

func TestWebhooks_SubscribeWebhookAuthError(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	// SDK retries once on 401 after a refresh attempt; queue a 401 for both.
	srv.Enqueue(testsupport.MockResponse{Status: 401, Body: `{"message":"bad token"}`,
		Headers: map[string]string{"Content-Type": "application/json"}})
	srv.Enqueue(testsupport.MockResponse{Status: 401, Body: `{"message":"refresh failed"}`,
		Headers: map[string]string{"Content-Type": "application/json"}})

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at", RefreshToken: "rt"})

	_, err := c.Webhooks().SubscribeWebhook(context.Background(), &mobiscroll.SubscribeWebhookParams{
		Provider:   mobiscroll.ProviderGoogle,
		CalendarID: "primary",
	})
	var ae *mobiscroll.AuthenticationError
	if !errors.As(err, &ae) {
		t.Fatalf("expected *AuthenticationError, got %T: %v", err, err)
	}
}

func TestWebhooks_UnsubscribeWebhookSendsJSONBodyAndParsesResponse(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.EnqueueJSON(`{"success": true}`)

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at"})

	resp, err := c.Webhooks().UnsubscribeWebhook(context.Background(), &mobiscroll.UnsubscribeWebhookParams{
		Provider:   mobiscroll.ProviderGoogle,
		ChannelID:  "chan1",
		ResourceID: "res1",
	})
	if err != nil {
		t.Fatalf("UnsubscribeWebhook: %v", err)
	}
	if !resp.Success {
		t.Fatalf("unexpected response: %+v", resp)
	}

	req := srv.Requests()[0]
	if req.Method != "POST" || req.Path != "/api/unsubscribe-webhook" {
		t.Fatalf("unexpected request: %s %s", req.Method, req.Path)
	}
	body := string(req.Body)
	for _, want := range []string{`"provider":"google"`, `"channelId":"chan1"`, `"resourceId":"res1"`} {
		if !strings.Contains(body, want) {
			t.Errorf("body missing %q: %s", want, body)
		}
	}
}

// TestWebhooks_UnsubscribeWebhookSuccessDespiteMessage covers the documented
// case where the backend already removed the local mapping and reports
// success even though the provider-side unsubscribe itself failed (e.g. an
// already-expired subscription). Callers should treat 200 as final.
func TestWebhooks_UnsubscribeWebhookSuccessDespiteMessage(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.EnqueueJSON(`{"success": true, "message": "provider subscription already expired"}`)

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at"})

	resp, err := c.Webhooks().UnsubscribeWebhook(context.Background(), &mobiscroll.UnsubscribeWebhookParams{
		Provider:  mobiscroll.ProviderGoogle,
		ChannelID: "chan1",
	})
	if err != nil {
		t.Fatalf("UnsubscribeWebhook: %v", err)
	}
	if !resp.Success || resp.Message == "" {
		t.Fatalf("expected success=true with an explanatory message, got %+v", resp)
	}
}

func TestWebhooks_UnsubscribeWebhookNilParams(t *testing.T) {
	c := mobiscroll.NewClient("id", "secret", "https://app/cb")
	if _, err := c.Webhooks().UnsubscribeWebhook(context.Background(), nil); err == nil {
		t.Fatal("expected error for nil params")
	}
}

func TestWebhooks_UnsubscribeWebhookValidationError(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.Enqueue(testsupport.MockResponse{
		Status:  400,
		Body:    `{"message":"channelId is required"}`,
		Headers: map[string]string{"Content-Type": "application/json"},
	})

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at"})

	_, err := c.Webhooks().UnsubscribeWebhook(context.Background(), &mobiscroll.UnsubscribeWebhookParams{
		Provider: mobiscroll.ProviderGoogle,
	})
	var ve *mobiscroll.ValidationError
	if !errors.As(err, &ve) {
		t.Fatalf("expected *ValidationError, got %T: %v", err, err)
	}
}

func TestWebhooks_UnsubscribeWebhookAuthError(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.Enqueue(testsupport.MockResponse{Status: 401, Body: `{"message":"bad token"}`,
		Headers: map[string]string{"Content-Type": "application/json"}})
	srv.Enqueue(testsupport.MockResponse{Status: 401, Body: `{"message":"refresh failed"}`,
		Headers: map[string]string{"Content-Type": "application/json"}})

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at", RefreshToken: "rt"})

	_, err := c.Webhooks().UnsubscribeWebhook(context.Background(), &mobiscroll.UnsubscribeWebhookParams{
		Provider:  mobiscroll.ProviderGoogle,
		ChannelID: "chan1",
	})
	var ae *mobiscroll.AuthenticationError
	if !errors.As(err, &ae) {
		t.Fatalf("expected *AuthenticationError, got %T: %v", err, err)
	}
}
